pkg/webhook/pod/pods.go

```diff
@@ -0,0 +1,169 @@
+// limitations under the License.
+
+package pod
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	memberUtils "github.com/pingcap/tidb-operator/pkg/manager/member"
+	"github.com/pingcap/tidb-operator/pkg/pdapi"
+	operatorUtils "github.com/pingcap/tidb-operator/pkg/util"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+type PodAdmissionControl struct {
+	// kubernetes client interface
+	kubeCli kubernetes.Interface
+	// operator client interface
+	operatorCli versioned.Interface
+	// pvc control
+	pvcControl controller.PVCControlInterface
+	// pd Control
+	pdControl pdapi.PDControlInterface
+	// pod Lister
+	podLister corelisters.PodLister
+}
+
+func NewPodAdmissionControl(kubeCli kubernetes.Interface, operatorCli versioned.Interface, PdControl pdapi.PDControlInterface, kubeInformerFactory kubeinformers.SharedInformerFactory, recorder record.EventRecorder) *PodAdmissionControl {
+
+	pvcInformer := kubeInformerFactory.Core().V1().PersistentVolumeClaims()
+	PVCControl := controller.NewRealPVCControl(kubeCli, recorder, pvcInformer.Lister())
+
+	podInformer := kubeInformerFactory.Core().V1().Pods()
+
+	return &PodAdmissionControl{
+		kubeCli:     kubeCli,
+		operatorCli: operatorCli,
+		pvcControl:  PVCControl,
+		pdControl:   PdControl,
+		podLister:   podInformer.Lister(),
+	}
+}
+
+func (pc *PodAdmissionControl) AdmitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+
+	name := ar.Request.Name
+	namespace := ar.Request.Namespace
+	operation := ar.Request.Operation
+	klog.Infof("receive admission to %s pod[%s/%s]", operation, namespace, name)
+
+	switch operation {
+	case v1beta1.Delete:
+		return pc.admitDeletePods(name, namespace)
+	default:
+		klog.Infof("Admit to %s pod[%s/%s]", operation, namespace, name)
+		return util.ARSuccess()
+	}
+}
+
+// Webhook server receive request to delete pod
+// if this pod wasn't member of tidbcluster, just let the request pass.
+// Otherwise, we check tidbcluster and statefulset which own this pod whether they were existed.
+// If either of them were deleted,we would let this request pass,
+//// otherwise we will check it decided by component type.
+func (pc *PodAdmissionControl) admitDeletePods(name, namespace string) *v1beta1.AdmissionResponse {
+
+	pod, err := pc.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		klog.Infof("failed to find pod[%s/%s] during delete it,admit to delete", namespace, name)
+		return util.ARSuccess()
+	}
+
+	l := label.Label(pod.Labels)
+	if !(l.IsPD() || l.IsTiKV() || l.IsTiDB()) {
```

Comment:
Only check PD for now?

Reply:
Yes.

Reply:
I mean why also check TiKV and TiDB here.